src/main/java/appeng/parts/misc/ItemHandlerAdapter.java

```diff
@@ -0,0 +1,370 @@
+			ItemStack oldIS = cachedStacks[slot];
+			IAEItemStack oldAeIS = cachedAeStacks[slot];
+
+			ItemStack newIS = itemHandler.getStackInSlot( slot );
+
+			if( this.isDifferent( newIS, oldIS ) )
+			{
+				// Completely different item
+				cachedStacks[slot] = newIS;
+				cachedAeStacks[slot] = AEItemStack.create( newIS );
+
+				// If we had a stack previously in this slot, notify the newtork about its disappearance
+				if( oldAeIS != null )
+				{
+					oldAeIS.setStackSize( -oldAeIS.getStackSize() );
+					changes.add( oldAeIS );
+				}
+
+				// Notify the network about the new stack. Note that this is null if newIS was null
+				if( cachedAeStacks[slot] != null )
+				{
+					changes.add( cachedAeStacks[slot] );
+				}
+			}
+			else
+			{
+				// Still the same item, but amount might have changed
+				int newSize = ( newIS == null ? 0 : newIS.stackSize );
+				int diff = newSize - ( oldIS == null ? 0 : oldIS.stackSize );
+
+				IAEItemStack stack = ( oldAeIS == null ? AEItemStack.create( newIS ) : oldAeIS.copy() );
+				if( stack != null )
+				{
+					stack.setStackSize( newSize );
+				}
+
+				if( diff != 0 && stack != null )
+				{
+					cachedStacks[slot] = newIS;
+					cachedAeStacks[slot] = stack;
+
+					final IAEItemStack a = stack.copy();
+					a.setStackSize( diff );
+					changes.add( a );
+				}
+			}
+		}
+
+		// Handle cases where the number of slots actually is lower now than before
+		if( slots < cachedStacks.length )
+		{
+			for (int slot = slots; slot < cachedStacks.length; slot++ )
+			{
+				IAEItemStack aeStack = cachedAeStacks[slot];
+				if ( aeStack != null ) {
+					IAEItemStack a = aeStack.copy();
+					a.setStackSize( -a.getStackSize() );
+					changes.add( a );
+				}
+			}
+
+			// Reduce the cache size
+			cachedStacks = Arrays.copyOf( cachedStacks, slots );
+			cachedAeStacks = Arrays.copyOf( cachedAeStacks, slots );
+		}
+
+		if( !changes.isEmpty() )
+		{
+			this.postDifference( changes );
+			return TickRateModulation.URGENT;
+		}
+		else
+		{
+			return TickRateModulation.SLOWER;
+		}
+	}
+
+	private boolean isDifferent( final ItemStack a, final ItemStack b )
+	{
+		if( a == b && b == null )
+		{
+			return false;
+		}
+
+		if( ( a == null && b != null ) || ( a != null && b == null ) )
+		{
+			return true;
+		}
+
+		return !Platform.isSameItemPrecise( a, b );
+	}
+
+	private void postDifference( Iterable<IAEItemStack> a )
+	{
+		final Iterator<Map.Entry<IMEMonitorHandlerReceiver<IAEItemStack>, Object>> i = this.listeners.entrySet().iterator();
+		while( i.hasNext() )
+		{
+			final Map.Entry<IMEMonitorHandlerReceiver<IAEItemStack>, Object> l = i.next();
+			final IMEMonitorHandlerReceiver<IAEItemStack> key = l.getKey();
+			if( key.isValid( l.getValue() ) )
+			{
+				key.postChange( this, a, mySource );
+			}
+			else
+			{
+				i.remove();
+			}
+		}
+	}
+
+	private BaseActionSource getActionSource()
+	{
+		return this.mySource;
+	}
+
+	public void setActionSource( final BaseActionSource mySource )
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Add the "@Override" annotation above this method signature [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.orod.org/coding_rules#rule_key=squid%3AS1161)